scality_manila_utils/smb_helper.py

```diff
@@ -0,0 +1,359 @@
+    """
+    @functools.wraps(f)
+    def wrapper(export_name, *args, **kwargs):
+        exports = _get_defined_exports()
+        if export_name not in exports:
+            msg = "Share '{0:s}' not found in Samba registry.".format(
+                  export_name)
+            raise ExportNotFoundException(msg)
+
+        return f(export_name=export_name, exports=exports, *args, **kwargs)
+
+    return wrapper
+
+
+@ensure_environment
+@ensure_export_exists
+def get_export(export_name, exports, *args, **kwargs):
+    """
+    Retrieve client details of an export.
+
+    :param export_name: name of export
+    :type export_name: string (unicode)
+    :param exports: all the defined shares in the Samba registry
+    :type exports: dictionary
+    :returns: string with export client details in json format
+    """
+
+    export = exports[export_name]
+    clients = dict((host, ["rw"]) for host in export['hosts allow'].split())
+
+    return json.dumps(clients)
+
+
+@ensure_environment
+def add_export(root_export, export_name, *args, **kwargs):
+    """
+    Add an export.
+
+    :param root_export: SOFS directory which holds the export points exposed
+        through manila
+    :type root_export: string (unicode)
+    :param export_name: name of export to add
+    :type export_name: string (unicode)
+    """
+
+    if not export_name or '/' in export_name:
+        raise ExportException('Invalid export name')
+
+    export_point = os.path.join(root_export, export_name)
+
+    # Copied from Manila generic driver
+    create_cmd = [
+        'net', 'conf', 'addshare', export_name, export_point,
+        'writeable=y', 'guest_ok=y',
+    ]
+    parameters = {
+        'browseable': 'yes',
+        '\"create mask\"': '0755',
+        '\"hosts deny\"': '0.0.0.0/0',  # deny all by default
+        '\"hosts allow\"': '127.0.0.1',
+        '\"read only\"': 'no',
+    }
+    set_of_commands = [':', ]  # : is just placeholder
+    for param, value in parameters.items():
+        # These are combined in one list to run in one process
+        # instead of big chain of one action calls.
+        set_of_commands.extend(['&&', 'net', 'conf', 'setparm',
+                                export_name, param, value])
+
+    with utils.elevated_privileges():
+        try:
+            os.mkdir(export_point)
+        except OSError as exc:
+            if exc.errno != errno.EEXIST:
+                raise
+            else:
+                log.debug("The share/directory %s already exists on SOFS",
+                          export_name)
+                exports = _get_defined_exports()
+                if export_name in exports:
+                    msg = ("Share '{0:s}' already defined in Samba "
+                           "registry.".format(export_name))
+                    raise ExportAlreadyExists(msg)
+
+        os.chmod(export_point, 0o0777)
+
+        subprocess.check_call(create_cmd)
+        subprocess.check_call(' '.join(set_of_commands), shell=True)
```

Comment:
`check_call` takes any `subprocess.Popen` call, which actually expects its first argument to be a list of strings, instead of a single string, unless you go through a shell (in which case this string will be passed to the shell after `-c`, and then expanded into a list).

So, instead of turning the `set_of_commands` list into a space-separated list, just keep it as a list. No need to escape any arguments, even.

It's worthwhile to spend some time reading the man-pages of `execvp` and others in that family of calls.

Obviously when using this (preferred) method (it's always better *not* to go through a shell, because that brings lots of pain potentially), you'll need to run all commands separately (which is also preferable: then at least you know which of them failed...), instead of the current concatenated version.